hazelcast/src/main/java/com/hazelcast/config/OnJoinCacheConfig.java

```diff
@@ -0,0 +1,163 @@
+     */
+    public OnJoinCacheConfig(CacheConfig cacheConfig) {
+        this(cacheConfig, true);
+    }
+
+    @SuppressWarnings("checkstyle:executablestatementcount")
+    public OnJoinCacheConfig(CacheConfig cacheConfig, boolean resolved) {
+        setAsyncBackupCount(cacheConfig.getAsyncBackupCount());
+        setBackupCount(cacheConfig.getBackupCount());
+        setCacheLoaderFactory(cacheConfig.getCacheLoaderFactory());
+        setCacheWriterFactory(cacheConfig.getCacheWriterFactory());
+        setDisablePerEntryInvalidationEvents(cacheConfig.isDisablePerEntryInvalidationEvents());
+        setEvictionConfig(cacheConfig.getEvictionConfig());
+        setExpiryPolicyFactory(cacheConfig.getExpiryPolicyFactory());
+        setHotRestartConfig(cacheConfig.getHotRestartConfig());
+        setInMemoryFormat(cacheConfig.getInMemoryFormat());
+        if (resolved) {
+            setKeyType(cacheConfig.getKeyType());
+            setValueType(cacheConfig.getValueType());
+        } else {
+            setKeyClassName(cacheConfig.getKeyClassName());
+            setValueClassName(cacheConfig.getValueClassName());
+        }
+        setManagementEnabled(cacheConfig.isManagementEnabled());
+        setManagerPrefix(cacheConfig.getManagerPrefix());
+        setMergePolicy(cacheConfig.getMergePolicy());
+        setName(cacheConfig.getName());
+        setPartitionLostListenerConfigs(cacheConfig.getPartitionLostListenerConfigs());
+        setQuorumName(cacheConfig.getQuorumName());
+        setReadThrough(cacheConfig.isReadThrough());
+        setStatisticsEnabled(cacheConfig.isStatisticsEnabled());
+        setStoreByValue(cacheConfig.isStoreByValue());
+        setUriString(cacheConfig.getUriString());
+        setWanReplicationRef(cacheConfig.getWanReplicationRef());
+        setWriteThrough(cacheConfig.isWriteThrough());
+        Iterable<CacheEntryListenerConfiguration<K, V>> entryListenerConfigs = cacheConfig.getCacheEntryListenerConfigurations();
+        for (CacheEntryListenerConfiguration<K, V> entryListenerConfig : entryListenerConfigs) {
+            addCacheEntryListenerConfiguration(entryListenerConfig);
+        }
+    }
+
+    @Override
+    void writeKeyValueTypes(ObjectDataOutput out)
+            throws IOException {
+        out.writeUTF(getKeyClassName());
+        out.writeUTF(getValueClassName());
+    }
+
+    @Override
+    void readKeyValueTypes(ObjectDataInput in)
+            throws IOException {
+        setKeyClassName(in.readUTF());
+        setValueClassName(in.readUTF());
+    }
+
+    @Override
+    public int getFactoryId() {
+        return ConfigDataSerializerHook.F_ID;
+    }
+
+    @Override
+    public int getId() {
+        return ConfigDataSerializerHook.ON_JOIN_CACHE_CONFIG;
+    }
+
+    /**
+     * @return this configuration as a {@link CacheConfig}
+     */
+    public CacheConfig asCacheConfig() {
+        CacheConfig<K, V> cacheConfig = new CacheConfig<K, V>();
```

Comment:
Can't you use `return new CacheConfig<K, V>(this)` to avoid duplicating all these setter calls?

Reply:
The `CacheConfig(CompleteConfiguration)` constructor will immediately resolve `keyType` & `valueType`. What we need here is to only set the key/value class names in the transformed-to `CacheConfig`, so the actual classes will be loaded later.